### PR TITLE
Removed some code duplication and added more error checking

### PR DESCRIPTION
**Pull request by gmbeard**

### CODE REVIEW

File: src/ip/mod.rs [hunk 1]
```diff
@@ -8,41 +12,26 @@ type NetworkValues = (NetworkAddr<u32>, BroadcastAddr<u32>);
...
-    let mut mask: u32 = 0;
-    for byte in octets {
-        let ibyte = byte.parse::<u8>().unwrap();
-        mask += 256u32.pow(octet_count as u32) * ibyte as u32;
-        if octet_count > 0 {
-            octet_count -= 1;
-        }
-    }
-    return mask
+pub fn make_mask_from_string(address: &str) -> Result<u32, ParseIntError> {
```
Comment (AndyGauge) [inline src/ip/mod.rs:26]: So this is a lot more idiomatic, but I've never seen some of these constructs before.  I need to understand what is happening here before I merge it so let me do research on `.enumerate()`, `.fold()`.  I'm assuming this is a map/reduce construct so I'm researching it.

File: src/ip/mod.rs [hunk 2]
```diff
@@ -8,41 +12,26 @@ type NetworkValues = (NetworkAddr<u32>, BroadcastAddr<u32>);
...
+    address.split(".")
+        .enumerate()
+        .fold(Ok(0_u32), |acc, (i, octet)| {
```
Comment (AndyGauge) [inline src/ip/mod.rs:29]: I can't find good documentation.  Is this a lambda?  Is |acc, (i, octet)| declaring the variables used within the codeblock?

**Comment by gmbeard** [comment]
Yeah, there's quite a bit going on in that expression. I'll try and break it down...

    address.split(".")

This creates the initial iterator. You're already familiar with this.

    .enumerate()

This transforms the iterator from producing `&str`s, to producing `(usize, &str)` items. The `usize` is the index of each item. We use this to determine which octet we're dealing with later on in the expression.

    .fold(Ok(0_u32), |acc, (i, octet)| {

This is the function / combinator that "collects" the iterator's values into a single value. Your assumption is correct; it's basically a "reduce" operation seen in other languages. The 1st argument, `Ok(0_u32)`,  is the initial value of the reduction (in this case it's a `Result<u32, ParseIntError>` type, which is inferred from `make_mask_from_string`'s return type). The 2nd argument, `|acc, (i, octet)| {`, is a [closure](https://doc.rust-lang.org/book/closures.html), which is an anonymous function (or lambda) that is called for each item in the iterator. Its arguments are: `acc`, the result of the previous call, and `(i, octet)`, which is the iterator's current value (a `(usize, &str)` tuple type). The fold expression expects this closure to produce a new value for `acc` using the current iterator value.

    acc.and_then(|mask| {

Because the type of `acc` is `Result<u32, ParseIntError>`, we're saying that any call of `fold`'s closure could produce an error (as a result of parsing the octet). [The `and_then` function](https://doc.rust-lang.org/std/result/enum.Result.html#method.and_then) makes sure that it only calls its closure argument if it's previous value was `Ok`, otherwise it just returns `Err`. Effectively, this means that if we get an error parsing a particular octet then we won't attempt to process any further octets.

    let ibyte = try!(octet.parse::<u8>()) as u32;
    Ok(mask | ibyte << (8 * (std::mem::size_of_val(&mask) - 1 - i)))

This is the part that attempts to convert each `&str` in the split to a byte and bitwise ORs it with the previous value. Notice the `try!` macro around the parse. This will do an "early return" if an error occurs during `parse()`. Also note that we use `i` in the second expression, which is the octet index provided by the earler call to `.enumerate()`. Finally we wrap the result in `Ok( .. )` because, as I mentioned, our `fold` call expects a `Result< .. >` type.

I guess you could say that this is idiomatic Rust. IMO, it doesn't always make for the clearest syntax but Rust's type system almost encourages these kind of expressions.

**Comment by AndyGauge** [comment]
Thank you so much for the explanation.  That was very close to what I was thinking it was doing, but struggled to find the Rust syntax for references.  The extra effort you put in to not only explain this expression by expression but also link the references will go a long way in my understanding of this type system.  I've done very little work in static languages, and Rust is where I'm trying to learn type systems.  Again, thank you for the links and the extra effort!

**Comment by gmbeard** [comment]
No problem at all 👍 Glad I could help. This application has inspired me to rewrite some of our IP validation tools at work, so it's been very useful.

**Pull request merged**